myia/unity.py

```diff
@@ -0,0 +1,503 @@
+        return value in self.legal_values
+
+    def __repr__(self) -> str:
+        return f"RestrictedVar({self.tag}, {self.legal_values})"
+
+
+class FilterVar(Var[U]):
+    """Variable restricted to values that pass a filter function."""
+
+    __slots__ = ('filter',)
+
+    def __init__(self, tag: str, filter: FnFiltT[U]) -> None:
+        """Create a FilterVar."""
+        super().__init__(tag)
+        self.filter = filter
+
+    @classmethod
+    def _parse_args(cls, args: Tuple[Any, ...], kwargs: Dict[Any, Any]):
+        assert len(kwargs) == 0
+        assert len(args) == 2
+        assert isinstance(args[0], str)
+        return args
+
+    def matches(self, value: Value[U]) -> bool:
+        """Return True if the variable matches the value."""
+        if isinstance(value, RestrictedVar):
+            return all(self.filter(v) for v in value.legal_values)
+        if isinstance(value, FilterVar):
+            return self.filter == value.filter
+        return self.filter(value)
+
+    def __repr__(self) -> str:
+        return f"FilterVar({self.tag}, {self.filter})"
+
+
+def expandlist(lst: Iterable[Value[U]]) -> List[Value[U]]:
+    """Flatten the Seq instances in a sequence."""
+    lst = list(lst)
+    off = 0
+    for i, e in enumerate(list(lst)):
+        if isinstance(e, Seq):
+            lst[off + i:off + i + 1] = e
+            off += len(e) - 1
+    return lst
+
+
+def noseq(fn: Callable[[Value[U]], Value[U]], u: Value[U]) -> Value[U]:
+    """Make sure that there are no Seq in the value."""
+    um = fn(u)
+    if isinstance(um, Seq):
+        raise TypeError("Multiple values in single-value position")
+    return um
+
+
+class Unification(Generic[T]):
+    """Unification engine."""
+
+    def __init__(self) -> None:
+        """Create a unification engine."""
+        self._seq = 1
+        self._vars: Dict[Any, Var[T]] = dict()
+
+    def _get_var(self, cls: Type[Var[T]], args: Tuple[Any, ...]) -> Var[T]:
+        pargs = cls._parse_args(args, dict())
+        if pargs not in self._vars:
+            self._vars[pargs] = cls(*pargs)
+        return self._vars[pargs]
+
+    def _get_tag(self, tag: str):
+        if tag is None:
+            tag = '_{}'.format(self._seq)
+            self._seq += 1
+        else:
+            if tag.startswith('_'):
+                raise ValueError('tag starting with _')
+        return tag
+
+    def var(self, tag: str = None,
+            filter: FilterT[T] = None)-> Var[T]:
+        """Create a variable for unification purposes.
+
+        Arguments:
+            tag: An identifier for the variable. Two variables with the
+                same filter and identifier will return the same object.
+            filter: A predicate, or a set of values the variable is
+                allowed to take.
+        """
+        tag = self._get_tag(tag)
+        if callable(filter):
+            return self._get_var(FilterVar, (tag, filter))
+        elif filter:
```

Comment:
This will fall through if `filter == ()` and create a `Var` which will match anything, instead of the correct behavior in this case which is to match nothing.

Reply:
ok